### PR TITLE
Add tests and refactor to use Nested annotation

### PR DESCRIPTION
**Pull request by julia-zack**
- Add tests for `getActiveFederationRedeemScript` and `getActiveFederationAddress` methods.
- Use _Nested_ to avoid duplication code and set common environments when needed/possible.

### CODE REVIEW

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java [hunk 1]
```diff
@@ -19,164 +19,332 @@
...
-    void getActiveFederation_withNullFederations_returnsGenesisFederation() {
-        Federation genesisFederation = FederationTestUtils.getGenesisFederation(federationMainnetConstants);
-        Federation activeFederation = federationSupport.getActiveFederation();
+    @Nested
+    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
+    @Tag("getActiveFederationTestsWithNullFederations")
+    class ActiveFederationTestsWithNullFederations {
+        @Tag("getActiveFederationTestsWithNullFederations")
+
+        @BeforeAll
```
Comment (julia-zack) [inline rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:79]: in each `TestInstance` we have to set the variables we wanna use, since [this](https://github.com/rsksmart/rskj/pull/2469/files#diff-835afc919a7b1b320019bf2bbba298e41cb0a9f11b4e60662b034be2762f683bR63) setUp is not being performed before.

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java [hunk 2]
```diff
@@ -19,164 +19,332 @@
...
+    private final Federation genesisFederation = FederationTestUtils.getGenesisFederation(federationMainnetConstants);
+    private Federation oldFederation;
+    private Federation newFederation;
     private FederationStorageProvider storageProvider;
-    private FederationSupportBuilder federationSupportBuilder;
+    private FederationSupportBuilder federationSupportBuilder = new FederationSupportBuilder();
+    private FederationSupport federationSupport;
 
     @BeforeEach
     void setUp() {
```
Comment (julia-zack) [inline rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:63]: this setUp could be removed if we end up having all different test instances

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java [hunk 3]
```diff
@@ -19,164 +19,332 @@
...
             .withFederationStorageProvider(storageProvider)
             .build();
     }
 
-    @Test
-    void getActiveFederation_withNullFederations_returnsGenesisFederation() {
-        Federation genesisFederation = FederationTestUtils.getGenesisFederation(federationMainnetConstants);
-        Federation activeFederation = federationSupport.getActiveFederation();
+    @Nested
+    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
```
Comment (julia-zack) [inline rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:74]: this was needed to can have parameterized tests inside nested annotation

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java [hunk 4]
```diff
@@ -19,164 +19,332 @@
...
+            Federation activeFederation = federationSupport.getActiveFederation();
+            assertThat(activeFederation, is(genesisFederation));
+        }
 
-    @Test
-    void getActiveFederation_withOldFederationAndNullNewFederation_returnsGenesisFederation() {
-        Federation oldFederation = FederationTestUtils.getErpFederation(networkParameters);
-        storageProvider.setOldFederation(oldFederation);
+        @Test
+        void getActiveFederationRedeemScript_beforeRSKIP293_returnsEmpty() {
```
Comment (julia-zack) [inline rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:98]: this was added too to prove this happens also in this context (not only when having two null feds)

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java [hunk 5]
```diff
@@ -19,164 +19,332 @@
...
+            );
+        }
 
-        // Assert
-        if (newFedExpectedToBeActive){
-            assertThat(activeFederation, is(newFederation));
-        } else {
-            assertThat(activeFederation, is(oldFederation));
+        @Test
+        void getActiveFederationRedeemScript_beforeRSKIP293_returnsEmpty() {
```
Comment (julia-zack) [inline rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:289]: again, this was added here too to prove this happens also in this context (not only when having two null feds)

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java [hunk 6]
```diff
@@ -19,164 +19,332 @@
...
+                .withFederationStorageProvider(storageProvider)
+                .withRskExecutionBlock(executionBlock)
+                .withActivations(activations)
+                .build();
+
+            Federation activeFederation = federationSupport.getActiveFederation();
+            assertThat(activeFederation, is(expectedFederation));
+        }
+        private Stream<Arguments> expectedFederation_args() {
+            return Stream.of(
```
Comment (marcos-iov) [inline rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:280]: A few more cases could be covered here. `blockNumberFederationActivationHop` with `fingerrootActivations` for example.

Assert that in fingerroot, it takes longet to activate the federation than in hop

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java [hunk 7]
```diff
@@ -19,164 +19,332 @@
...
+                Arguments.of(blockNumberFederationActivationHop - 1, oldFederation, hopActivations), // new federation shouldn't be active
+                Arguments.of(blockNumberFederationActivationHop + 1, newFederation, hopActivations), // new federation should be active
+                Arguments.of(blockNumberFederationActivationFingerroot - 1, oldFederation, fingerrootActivations), // new federation shouldn't be active
+                Arguments.of(blockNumberFederationActivationFingerroot + 1, newFederation, fingerrootActivations) // new federation should be active
```
Comment (marcos-iov) [inline rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:284]: Maybe check using the exact activation block number? And not +1, to be sure we cover that corner case

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java [hunk 8]
```diff
@@ -19,164 +19,332 @@
...
+            federationSupport = federationSupportBuilder
+                .withFederationConstants(federationMainnetConstants)
+                .withFederationStorageProvider(storageProvider)
+                .build();
+            assertFalse(federationSupport.getActiveFederationRedeemScript().isPresent());
+        }
+
+        @ParameterizedTest
+        @MethodSource("expectedRedeemScript_args")
+        void getActiveFederationRedeemScript_returnsExpectedRedeemScriptAccordingToActivationAgeAndActivations(long currentBlock, Script expectedRedeemScript,
```
Comment (marcos-iov) [inline rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:299]: Here and in other tests, I'd put the expected as the last param. Makes it easier to understand for me at least. First the setup values then finally the expected

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationTestUtils.java [hunk 9]
```diff
@@ -280,4 +279,4 @@ private static Script createInputScriptSig(
             .data(fedRedeemScript.getProgram())
             .build();
     }
-}
+}
```
Comment (marcos-iov) [inline rskj-core/src/test/java/co/rsk/peg/federation/FederationTestUtils.java:282]: Don't remove please :)

Comment (julia-zack) [reply rskj-core/src/test/java/co/rsk/peg/federation/FederationTestUtils.java:282]: i thought i had added the blank line, not removed it. My bad!

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java [hunk 10]
```diff
@@ -19,164 +19,332 @@
...
-            newFedCreationBlockNumber
-        );
+            Optional<Script> activeFederationRedeemScript = federationSupport.getActiveFederationRedeemScript();
+            assertTrue(activeFederationRedeemScript.isPresent());
+            assertThat(activeFederationRedeemScript.get(), is(expectedRedeemScript));
+        }
+        private Stream<Arguments> expectedRedeemScript_args() {
+            return Stream.of(
+                Arguments.of(oldFederation, null, genesisFederation.getRedeemScript()),
+                Arguments.of(null, newFederation, newFederation.getRedeemScript())
```
[the remark this answers is not in the document]
Comment (julia-zack) [reply rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:197]: nope, just when we have both federations and the new is not active yet. Check the [implementation](https://github.com/rsksmart/rskj/blob/38fe20bb541785e96c8843fe70de506ee890e8e5/rskj-core/src/main/java/co/rsk/peg/federation/FederationSupportImpl.java#L105)

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java [hunk 11]
```diff
@@ -19,164 +19,332 @@
...
+            storageProvider.setOldFederation(oldFederation);
+            storageProvider.setNewFederation(newFederation);
+
+            Address activeFederationAddress = federationSupport.getActiveFederationAddress();
+            assertThat(activeFederationAddress, is(expectedAddress));
+        }
+        private Stream<Arguments> expectedAddress_args() {
+            return Stream.of(
+                Arguments.of(oldFederation, null, genesisFederation.getAddress()),
+                Arguments.of(null, newFederation, newFederation.getAddress())
```
[the remark this answers is not in the document]
Comment (julia-zack) [reply rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:213]: same answer here :)

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java (hunk 6 above)
Comment (julia-zack) [reply rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:280]: good one, thanks :)

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java [hunk 12]
```diff
@@ -19,164 +19,332 @@
...
-            .build();
+            federationSupport = federationSupportBuilder
+                .withFederationConstants(federationMainnetConstants)
+                .withFederationStorageProvider(storageProvider)
+                .withActivations(activations)
+                .build();
 
-        networkParameters = federationConstants.getBtcParams();
+            storageProvider.setOldFederation(oldFederation);
+            storageProvider.setNewFederation(newFederation);
```
[the remark this answers is not in the document]
Comment (julia-zack) [reply rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:188]: is the same, in real life we are doing it this way right?

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java (hunk 6 above)
Comment (julia-zack) [reply rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:300]: added!

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java (hunk 7 above)
Comment (julia-zack) [reply rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java]: done

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (marcos-iov) [reply rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:216]: Ok, should we have cases for that scenario?

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (marcos-iov) [reply rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:206]: Yeah, it actually works the same because storageProvider is passed as a reference to FederationSupport

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (julia-zack) [reply rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:216]: we have! [These ones](https://github.com/rsksmart/rskj/blob/feature/tests/getActiveFederationAddress/rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java#L239) are those tests.
Check the expected [here](https://github.com/rsksmart/rskj/blob/feature/tests/getActiveFederationAddress/rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java#L301) and [here](https://github.com/rsksmart/rskj/blob/feature/tests/getActiveFederationAddress/rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java#L304)

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java [hunk 13]
```diff
@@ -19,164 +19,362 @@
...
+            assertThat(activeFederationRedeemScript.get(), is(expectedRedeemScript));
+        }
 
-        storageProvider.setNewFederation(newFederation);
-        storageProvider.setOldFederation(oldFederation);
+        private Stream<Arguments> expectedRedeemScript_args() {
+            return Stream.of(
+                Arguments.of(blockNumberFederationActivationHop - 1, hopActivations, oldFederation.getRedeemScript()), // new federation shouldn't be active
+                Arguments.of(blockNumberFederationActivationHop, hopActivations, newFederation.getRedeemScript()), // new federation should be active
+                Arguments.of(blockNumberFederationActivationHop, fingerrootActivations, oldFederation.getRedeemScript()), // new federation shouldn't be active since hop activation block number is smaller than fingerroot one
```
Comment (apancorb) [inline rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:345]: nit: since this comment is very long, consider putting it above instead of next to it

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java [hunk 14]
```diff
@@ -19,164 +19,362 @@
...
-    private FederationSupport federationSupport;
-    private FederationConstants federationMainnetConstants;
-    private NetworkParameters networkParameters;
+    private static final FederationConstants federationMainnetConstants = FederationMainNetConstants.getInstance();
+    private final Federation genesisFederation = FederationTestUtils.getGenesisFederation(federationMainnetConstants);
+    private Federation oldFederation;
+    private Federation newFederation;
     private FederationStorageProvider storageProvider;
-    private FederationSupportBuilder federationSupportBuilder;
+    private FederationSupportBuilder federationSupportBuilder = new FederationSupportBuilder();
```
Comment (apancorb) [inline rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:59]: If `federationSupportBuilder` is being initialized in the setUp phases, there shouldn't be a need to initialize it here.

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java (hunk 2 above)
Comment (apancorb) [reply rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:63]: Is this setUp method being called per test? Why do we need it? If there is no use, consider removing it.

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java [hunk 15]
```diff
@@ -19,164 +19,362 @@
...
 
-    @Test
-    void getActiveFederation_withNullFederations_returnsGenesisFederation() {
-        Federation genesisFederation = FederationTestUtils.getGenesisFederation(federationMainnetConstants);
-        Federation activeFederation = federationSupport.getActiveFederation();
+    @Nested
+    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
+    @Tag("getActiveFederationTestsWithNullFederations")
+    class ActiveFederationTestsWithNullFederations {
+        @Tag("getActiveFederationTestsWithNullFederations")
```
[the remark this answers is not in the document]
Comment (julia-zack) [reply rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:77]: you right, i messed up some things. Thanks!

**Review by apancorb — APPROVED** [review]
LGTM

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java [hunk 16]
```diff
@@ -19,164 +19,371 @@
...
+            storageProvider.setNewFederation(newFederation);
 
-        Federation activeFederation = federationSupport.getActiveFederation();
+            Federation activeFederation = federationSupport.getActiveFederation();
+            assertThat(activeFederation, is(expectedFederation));
+        }
+
+        private Stream<Arguments> expectedFederationArgs() {
+            return Stream.of(
+                Arguments.of(oldFederation, null, genesisFederation),
```
Comment (nathanieliov) [inline rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:174]: Is this a real scenario? Is it possible to have an old federation and an active federation to be the genesis? That's not realistic. Once you have an old federation, one federation must be active, different than the genesis, because for an old federation exists, the genesis should be changed.

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:217]: As I mentioned above, this scenario is not realistic and is confusing to me. According to the implementation, the genesis federation should not be used anymore once an old federation exists.

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:235]: Same argument as above.

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java [hunk 17]
```diff
@@ -19,164 +19,371 @@
...
+
         ActivationConfig.ForBlock hopActivations = ActivationConfigsForTest.hop400().forBlock(0);
+        long blockNumberFederationActivationHop;
         ActivationConfig.ForBlock fingerrootActivations = ActivationConfigsForTest.fingerroot500().forBlock(0);
+        long blockNumberFederationActivationFingerroot;
+        FederationStorageProvider storageProvider;
+
+        @BeforeAll
+        void setUp() {
+            long oldFederationCreationBlockNumber = 0;
```
Comment (nathanieliov) [inline rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:256]: What do you think about setting a value greater than zero, like 200? Since this is not the genesis, it is not possible for it to be zero. Just to be more realistic.

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java [hunk 18]
```diff
@@ -19,164 +19,371 @@
...
 
-        networkParameters = federationConstants.getBtcParams();
+        private Stream<Arguments> expectedFederationArgs() {
+            return Stream.of(
+                Arguments.of(blockNumberFederationActivationHop - 1, hopActivations, oldFederation),
+                Arguments.of(blockNumberFederationActivationHop, hopActivations, newFederation),
+                Arguments.of(blockNumberFederationActivationHop, fingerrootActivations, oldFederation),
+                Arguments.of(blockNumberFederationActivationFingerroot - 1, fingerrootActivations, oldFederation),
+                Arguments.of(blockNumberFederationActivationFingerroot, fingerrootActivations, newFederation),
+                Arguments.of(blockNumberFederationActivationFingerroot, hopActivations, newFederation)
```
Comment (nathanieliov) [inline rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:312]: Well done. Each case is clear and easy to follow.

**Review by marcos-iov — APPROVED** [review]

**Review by jeremy-then — APPROVED** [review]
Besides the suggested changes, LGTM.

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java (hunk 17 above)
Comment (julia-zack) [reply rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:256]: good one, thanks :)

**Review by nathanieliov — APPROVED** [review]
